### PR TITLE
[Foundation] Correct data subscript indexing to be offset from the base index

### PR DESCRIPTION
**Pull request by phausler**
Resolves https://bugs.swift.org/browse/SR-4964

Subsequences of Data inappropriately ignored the slice regions of both bridged data as well as certain index accessors.

### CODE REVIEW

**Comment by phausler** [comment]
@swift-ci please test

**Comment by phausler** [comment]
@swift-ci please test

**Comment by phausler** [comment]
@swift-ci please test

**Comment by phausler** [comment]
@swift-ci please test

**Comment by phausler** [comment]
@swift-ci please test

File: test/stdlib/TestData.swift [hunk 1]
```diff
@@ -739,7 +739,7 @@ class TestData : TestDataSuper {
         data.append("test", count: 4)
         expectTrue(object.verifier.wasMutableCopied)
 
-        let preservedObjectness = (data as NSData) is MutableDataVerifier
+        let preservedObjectness = (data as NSData) is ImmutableDataVerifier
```
Comment (phausler) [inline test/stdlib/TestData.swift]: this should never have been mutable since it would leak out potentially unsafe shared changes out from under the storage

File: test/stdlib/TestData.swift [hunk 2]
```diff
@@ -759,7 +759,7 @@ class TestData : TestDataSuper {
         data.append("test", count: 4)
         expectTrue(object.verifier.wasMutableCopied)
         
-        let preservedObjectness = (data as NSData) is MutableDataVerifier
+        let preservedObjectness = (data as NSData) is ImmutableDataVerifier
```
Comment (phausler) [inline test/stdlib/TestData.swift]: ditto here on leaking unsafe state

File: stdlib/public/SDK/Foundation/Data.swift [hunk 3]
```diff
@@ -11,32 +11,32 @@
 //===----------------------------------------------------------------------===//
 
 #if DEPLOYMENT_RUNTIME_SWIFT
-
+
```
Comment (michael-lehew) [inline stdlib/public/SDK/Foundation/Data.swift]: do we care at all about extraneous whitespace?

File: stdlib/public/SDK/Foundation/Data.swift [hunk 4]
```diff
@@ -703,50 +703,64 @@ public final class _DataStorage {
...
+            return try work(d)
         case .customMutableReference(let d):
-            if range.lowerBound == 0 && range.upperBound == _length {
-                return try work(d)
-            } else {
-                return try work(d.subdata(with: NSRange(location: range.lowerBound, length: range.count))._bridgeToObjectiveC())
+            guard range.lowerBound == 0 && range.upperBound == d.length else {
+                return try work(NSData(bytesNoCopy: UnsafeMutableRawPointer(mutating: d.bytes.advanced(by: range.lowerBound)), length: range.count, freeWhenDone: false))
             }
+            return try work(d)
```
Comment (michael-lehew) [inline stdlib/public/SDK/Foundation/Data.swift]: this looks correct, stylistically the multiple returns and repeated work invocations is a bit hard to follow.  it seems the branching is about getting a different 'd: NSData' in certain cases;

File: stdlib/public/SDK/Foundation/Data.swift (hunk 3 above)
Comment (phausler) [inline stdlib/public/SDK/Foundation/Data.swift]: It is from me moving it to and from Xcode to test in swift-corelibs-foundation; ideally we should have a consistent story of copy-paste structural types to and from SCLF
So it is not a problem imho.

File: stdlib/public/SDK/Foundation/Data.swift (hunk 4 above)
Comment (phausler) [inline stdlib/public/SDK/Foundation/Data.swift]: each associated type is a different storage to the backing. some are NSData where as some are NSMutableData and others are known to be a subclass of NSData or a known subclass of NSMutableData. I guess it could be a clarity enhancement to perhaps extract this as it's own function.

File: stdlib/public/SDK/Foundation/Data.swift (hunk 3 above)
Comment (michael-lehew) [inline stdlib/public/SDK/Foundation/Data.swift]: works for me.

**Review by michael-lehew — APPROVED** [review]
LGTM

File: test/stdlib/TestData.swift [hunk 5]
```diff
@@ -1008,6 +1008,58 @@ class TestData : TestDataSuper {
...
+        expectEqual(expected, slice)
+    }
+    
+    func test_sliceEquality2() {
+        let d = Data(bytes: [5, 6, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12])
+        let slice1 = d[0..<2]
+        let slice2 = d[5..<7]
+        expectEqual(slice1, slice2)
+    }
+
```
Comment (phausler) [inline test/stdlib/TestData.swift]: slices can be sliced; I am writing a few more tests to verify so I can add that

**Comment by phausler** [comment]
@swift-ci please test

**Comment by phausler** [comment]
@swift-ci please test

**Comment by phausler** [comment]
@swift-ci please test
